aten/src/ATen/CUDAStream.h

```diff
@@ -0,0 +1,102 @@
+#pragma once
+
+#include <cstdint>
+ 
+/*
+* A CPU-friendly CUDAStream interface.
+* 
+* Includes the CUDAStream RAII class and and internal stream API.
+* 
+* The ATen Context interface should be preferred when working with streams.
+*/
+
+// Forward-declares cudaStream_t to avoid depending on CUDA in CPU builds
+// Note: this is the internal CUDA runtime typedef for cudaStream_t
+struct CUstream_st;
+typedef struct CUstream_st* cudaStream_t;
+
+// Forward-declares internals
+struct CUDAStreamInternals;
+
+namespace at {
+
+// Pointer-based API (for internal use)
+// Consumers of streams should generally use ATen/Context to work with them safely
+CUDAStreamInternals* CUDAStream_getDefaultStreamOnDevice(int64_t device);
+CUDAStreamInternals* CUDAStream_getDefaultStream();
+
+CUDAStreamInternals* CUDAStream_createAndRetainWithOptions(int32_t flags, int32_t priority);
+
+CUDAStreamInternals* CUDAStream_getAndRetainCurrentStreamOnDevice(int64_t device);
+CUDAStreamInternals* CUDAStream_getAndRetainCurrentStream();
+
+// Note: these Unsafe gets should NEVER be used and are only here for legacy
+// purposes. Once those uses are gone they should be removed.
+CUDAStreamInternals* CUDAStream_getCurrentStreamOnDeviceUnsafe(int64_t device);
+CUDAStreamInternals* CUDAStream_getCurrentStreamUnsafe();
+
+void CUDAStream_setStreamOnDevice(int64_t device, CUDAStreamInternals* internals);
+void CUDAStream_setStream(CUDAStreamInternals* internals);
+
+cudaStream_t CUDAStream_stream(CUDAStreamInternals*);
+int CUDAStream_device(CUDAStreamInternals*);
+
+bool CUDAStream_retain(CUDAStreamInternals*);
+void CUDAStream_free(CUDAStreamInternals*&);
+
+CUDAStreamInternals* CUDAStream_copy(CUDAStreamInternals*);
+void CUDAStream_move(CUDAStreamInternals*& src, CUDAStreamInternals*& dst);
```

Comment:
It would be good to hide those from the plain `at` namespace (at least in `detail`), but generally why can't they just be out-of-class definitions of `CUDAStream` member function?

Reply:
If you're referring to only copy and move I totally agree (it's the 2nd bullet above). Happy to fix that up.